src/rendezvous/channels/ecdhV1.ts

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { SAS } from '@matrix-org/olm';
+
+import { RendezvousError } from '../error';
+import {
+    RendezvousCode,
+    RendezvousIntent,
+    RendezvousChannel,
+    RendezvousTransportDetails,
+    RendezvousTransport,
+    RendezvousFailureReason,
+} from '../index';
+import { SecureRendezvousChannelAlgorithm } from '.';
+import { encodeBase64, decodeBase64 } from '../../crypto/olmlib';
+import { getCrypto } from '../../utils';
+
+const subtleCrypto = (typeof window !== "undefined" && window.crypto) ?
+    (window.crypto.subtle || window.crypto.webkitSubtle) : null;
+
+export interface ECDHv1RendezvousCode extends RendezvousCode {
+    rendezvous: {
+        transport: RendezvousTransportDetails;
+        algorithm: SecureRendezvousChannelAlgorithm.ECDH_V1;
+        key: string;
+    };
+}
+
+// The underlying algorithm is the same as:
+// https://github.com/matrix-org/matrix-js-sdk/blob/75204d5cd04d67be100fca399f83b1a66ffb8118/src/crypto/verification/SAS.ts#L54-L68
+function generateDecimalSas(sasBytes: number[]): string {
+    /**
+     *      +--------+--------+--------+--------+--------+
+     *      | Byte 0 | Byte 1 | Byte 2 | Byte 3 | Byte 4 |
+     *      +--------+--------+--------+--------+--------+
+     * bits: 87654321 87654321 87654321 87654321 87654321
+     *       \____________/\_____________/\____________/
+     *         1st number    2nd number     3rd number
+     */
+    const digits = [
+        (sasBytes[0] << 5 | sasBytes[1] >> 3) + 1000,
+        ((sasBytes[1] & 0x7) << 10 | sasBytes[2] << 2 | sasBytes[3] >> 6) + 1000,
+        ((sasBytes[3] & 0x3f) << 7 | sasBytes[4] >> 1) + 1000,
+    ];
+
+    return digits.join('-');
+}
+
+async function importKey(key: Uint8Array): Promise<CryptoKey | Uint8Array> {
+    if (getCrypto()) {
```

Comment:
https://github.com/matrix-org/matrix-js-sdk/pull/2762 removes this function, so this may need to be updated (as well as the definition of `subtleCrypto`, and changing things in general to use only the WebCrypto API).  This will need some coordination between this PR and 2762.

Reply:
(The irony of having created more work for myself by suggesting the migration to Web Crypto this morning)

Reply:
Now rebased to use Web Crypto exclusively.